observer/observer_test.go

```diff
@@ -0,0 +1,141 @@
+	cmtjson "github.com/cometbft/cometbft/libs/json"
+	"github.com/cometbft/cometbft/libs/log"
+	coretypes "github.com/cometbft/cometbft/rpc/core/types"
+	"github.com/cometbft/cometbft/rpc/jsonrpc/types"
+	cmttypes "github.com/cometbft/cometbft/types"
+	proto "github.com/cosmos/gogoproto/proto"
+	"github.com/gorilla/websocket"
+	"github.com/stretchr/testify/require"
+
+	bridge "github.com/osmosis-labs/osmosis/v23/x/bridge/types"
+)
+
+const (
+	EventOutboundTransferType = "osmosis.bridge.v1beta1.EventOutboundTransfer"
+)
+
+var upgrader = websocket.Upgrader{}
+
+func readNewBlockEvent(path string) coretypes.ResultEvent {
+	dataStr, err := os.ReadFile(path)
+	if err != nil {
+		panic(err)
+	}
+	result := coretypes.ResultEvent{}
+	err = cmtjson.Unmarshal([]byte(dataStr), &result)
+	if err != nil {
+		panic(err)
+	}
+	return result
+}
+
+func readBlockResults(path string) coretypes.ResultBlockResults {
+	dataStr, err := os.ReadFile(path)
+	if err != nil {
+		panic(err)
+	}
+	result := coretypes.ResultBlockResults{}
+	err = json.Unmarshal([]byte(dataStr), &result)
+	if err != nil {
+		panic(err)
+	}
+	return result
+}
+
+func success(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "GET" {
+		c, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			panic(err)
+		}
+		defer c.Close()
+		newBlock := readNewBlockEvent("./test_events/new_block_event.json")
+		newBlockResp := types.NewRPCSuccessResponse(
+			types.JSONRPCIntID(1),
+			newBlock,
+		)
+		newBlockRaw, err := json.Marshal(newBlockResp)
+		if err != nil {
+			panic(err)
+		}
+
+		err = c.WriteMessage(1, newBlockRaw)
+		if err != nil {
+			panic(err)
+		}
+	} else if r.Method == "POST" {
+		blockResults := readBlockResults("./test_events/block_results.json")
+		blockResultsResp := types.NewRPCSuccessResponse(
+			types.JSONRPCIntID(0),
+			blockResults,
+		)
+		blockResultsRaw, err := json.Marshal(blockResultsResp)
+		if err != nil {
+			panic(err)
+		}
+		_, err = w.Write(blockResultsRaw)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+// Verify Observer properly reads subscribed messages and filters events
+func TestObserver(t *testing.T) {
+	s := httptest.NewServer(http.HandlerFunc(success))
+	defer s.Close()
+
+	observer, err := NewObserver(log.NewNopLogger(), s.URL)
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	query := cmttypes.QueryForEvent(cmttypes.EventNewBlock)
+	observeEvents := []string{proto.MessageName(&bridge.EventOutboundTransfer{})}
+
+	err = observer.Start(ctx, query.String(), observeEvents)
+	require.NoError(t, err)
+
+	// We expect Observer to receive 3 Txs with `EventOutboundTransferType` events in this test
+	// Only 3 of the Txs are successful, so we should receive only 2 event through the channel
+	eventsOut := observer.Events()
+	events := [2]abcitypes.Event{}
+	for i := 0; i < len(events); i++ {
+		select {
+		case e := <-eventsOut:
+			events[i] = e
+		case <-time.After(1 * time.Second):
+			panic("Channel read timeout")
+		}
```

Comment:
As I mentioned, panic is not a good idea in tests. We can replace panic with `t.Fatalf(timeout reading events from the observer)` But Fatal doesn't call defers, so it's not really correct to call it. The best scenario would be to use special time-based checks from the `require` package
```suggestion
		require.Eventually(t, func() bool {
			events[i] = <-eventsOut
			return true
		}, time.Second, 100*time.Millisecond,
		"timeout reading events from observed")
```
This call from time to time (every 100 ms) checks the execution and properly fails the test if the timeout (1 sec) is exceeded